feat(api): improve gripper error checking

sfoster1 opened the pull request:
When we move labware during the gripper, we use the position of the
labware jaw while labware is nominally gripped to determine whether we
successfully grabbed the labware or not. We compare the position to some
fact of the labware's geometry.

Because the geometry doesn't actually contain the width at the point
that it's gripped (it only has the bounding box, which includes e.g. a
skirt that the gripper doesn't interact with) and because the gripper's
jaw width position tracking can have problems, we also allow
uncertainty.

There were a couple problems that are now fixed, such as
- We were doing an exact position comparison for the
gripper-fully-closed case; this is not really safe because the encoder
can be a little off, so add a less-than check to make sure we detect it
- The big one: we determined our uncertainty values by the difference
between the well bounding box and the labware bounding box. For labware
that doesn't have wells, that means the uncertainty is effectively
infinite. Instead, use a fixed distance.
- Also, refactor some stuff to support it.

The behavior for things that have wells should be unchanged.

## testing
- see if we can detect failed lid pickups now!

Closes EXEC-1783

sfoster1 committed:
chore: log gripper position checking

sfoster1 committed:
feat(api): improve gripper error checking

When we move labware during the gripper, we use the position of the
labware jaw while labware is nominally gripped to determine whether we
successfully grabbed the labware or not. We compare the position to some
fact of the labware's geometry.

Because the geometry doesn't actually contain the width at the point
that it's gripped (it only has the bounding box, which includes e.g. a
skirt that the gripper doesn't interact with) and because the gripper's
jaw width position tracking can have problems, we also allow
uncertainty.

There were a couple problems that are now fixed, such as
- We were doing an exact position comparison for the
gripper-fully-closed case; this is not really safe because the encoder
can be a little off, so add a less-than check to make sure we detect it
- The big one: we determined our uncertainty values by the difference
between the well bounding box and the labware bounding box. For labware
that doesn't have wells, that means the uncertainty is effectively
infinite. Instead, use a fixed distance.
- Also, refactor some stuff to support it.

Closes EXEC-1783

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/labware.py:
```diff
@@ -1430,3 +1431,68 @@ def get_well_bbox(self, labware_definition: LabwareDefinition) -> Dimensions:
+    def _gripper_uncertainty_wider(
+        self, labware_bbox: Dimensions, well_bbox: Dimensions, target_grip_width: float
+    ) -> float:
+        """Most wider the gripper can be than the target while still likely gripping successfully.
+
+        This can be a lot closer to 0, since the bounding box of the labware will certainly be the
+        widest point (if it's defined without error), but since there might be error in the
+        definition we allow some slop.
+
+        The number will be interpreted relative to the target width, which is (for now) the labware
+        outer bounding box.
+
+        TODO: This should be a number looked up from the definition.
+        """
+        # This will be 0 unless the wells are wider than the labware
+        return max(well_bbox.y - target_grip_width, 0)
```
This `#` comment says that this will normally be 0, which looks true to me from the code. But the docstring says "since there might be error in the definition we allow some slop." Was there intended to be a hard-coded slop constant here, too?

sfoster1 replied:
No, the slop in this case is the wells-are-wider-than-the-labware part

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/labware.py:
```diff
@@ -1430,3 +1431,68 @@ def get_well_bbox(self, labware_definition: LabwareDefinition) -> Dimensions:
+    ) -> GripSpecs:
+        """Get the target and bounds for a successful grip of this labware."""
+        outer_bounds = self.get_dimensions(labware_definition=labware_definition)
+        well_bounds = self.get_well_bbox(labware_definition=labware_definition)
```
Not a problem introduced in this PR, but it looks like we're not correctly handling the case where the well bounds are not centered within the outer bounds, are we?

sfoster1 replied:
Yeah we definitely are not. This is why I want to just have these as numbers in the definitions

SyntaxColoring reviewed (approved):
Makes sense, thank you for the detailed comments. One point about a potentially missing slop constant, but otherwise LGTM.

sfoster1 committed:
bad number

sfoster1 committed:
more updates

I knew there was something wrong with the stupid ordering of the stupid variables
